### PR TITLE
Implement fullscreen modal with BS breakpoints

### PR DESCRIPTION
**Pull request by Ayoahha**

### CODE REVIEW

File: addon/components/bs-modal.js [hunk 1]
```diff
@@ -193,6 +193,20 @@ export default class Modal extends Component {
...
 
+    /**
+   *  Allows adding fullscreen mode for modals. It will
+   *  apply the `modal-fullscreen' class when using 'true' and
+   *  `modal-fullscreen-[x]-down` class when using BS breakpoints
+   *   ([x] = 'sm', 'md', 'lg', 'xl', 'xxl').
+   *  
+   * @property fullscreen
+   * @type {string}
+   * @default 'true'
```
Comment (jelhan) [inline addon/components/bs-modal.js:204]: It must not have a default value. Otherwise every modal would be fullscreen unless developer overwrites it explicitly. This would break backward compatibility.

File: addon/components/bs-modal.js [hunk 2]
```diff
@@ -267,7 +281,6 @@ export default class Modal extends Component {
    * @public
    */
   @arg
-  transitionDuration = 300;
```
Comment (jelhan) [inline addon/components/bs-modal.js:270]: I guess you had removed this line by accident?

File: addon/components/bs-modal.js [hunk 3]
```diff
@@ -278,7 +291,6 @@ export default class Modal extends Component {
    * @public
    */
   @arg
-  backdropTransitionDuration = 150;
```
Comment (jelhan) [inline addon/components/bs-modal.js:281]: As per above, I assume you had removed this line by accident?

File: addon/components/bs-modal/dialog.hbs [hunk 4]
```diff
@@ -13,9 +13,15 @@
...
-    <div class="modal-content"
-     tabindex="-1"
+  <div class="modal-dialog {{this.sizeClass}} {{if @centered "modal-dialog-centered"}} {{if @scrollable "modal-dialog-scrollable"}} 
+  {{#if @fullscreen}}
+    {{#if (bs-eq @fullscreen true)}}
+      modal-fullscreen
+    {{else}}
+      modal-fullscreen-{{@fullscreen}}-down
+    {{/if}}
+  {{/if}}" role="document">
```
Comment (jelhan) [inline addon/components/bs-modal/dialog.hbs:23]: If I recall correctly, we are currently trying to prevent linebreaks in `class` attributes even if that is causing very long lines.

```suggestion
  <div class="modal-dialog {{this.sizeClass}} {{if @centered "modal-dialog-centered"}} {{if @scrollable "modal-dialog-scrollable"}} {{#if @fullscreen}}{{#if (bs-eq @fullscreen true)}}modal-fullscreen{{else}}modal-fullscreen-{{@fullscreen}}-down{{/if}}{{/if}}" role="document">
```

File: addon/components/bs-modal/dialog.hbs [hunk 5]
```diff
@@ -13,9 +13,15 @@
...
+    <div class="modal-content" tabindex="-1"
```
Comment (jelhan) [inline addon/components/bs-modal/dialog.hbs:24]: Any specific reason for which you removed the linebreak before `tabindex`? Or was that just by mistake?

File: tests/integration/components/bs-modal-test.js [hunk 6]
```diff
@@ -312,5 +312,31 @@ module('Integration | Component | bs-modal', function (hooks) {
       await settled();
       assert.dom('.modal').doesNotExist();
     });
+
+    test('it passes along fullscreen class attribute', async function (assert) {
+      this.set('fullscreen', 'full');
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:317]: If I understand API documentation correctly, `full` is not a supported option. It should be `true` for this test case, isn't it?

```suggestion
      this.set('fullscreen', true);
```

File: tests/integration/components/bs-modal-test.js [hunk 7]
```diff
@@ -312,5 +312,31 @@ module('Integration | Component | bs-modal', function (hooks) {
       await settled();
       assert.dom('.modal').doesNotExist();
     });
+
+    test('it passes along fullscreen class attribute', async function (assert) {
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:316]: Small proposal to improve wording for clarity:

```suggestion
    test('it allows to render modal in fullscreen using @fullscreen argument', async function (assert) {
```

Comment (jelhan) [reply tests/integration/components/bs-modal-test.js:316]: Would be great if you could add a test that none of the classes is applied if `@fullscreen` argument is not passed.

Comment (jelhan) [reply tests/integration/components/bs-modal-test.js:316]: It seems that this test is currently grouped under the `it animates opening and closing the modal`. Would be great if you could move it above into the main module.

File: addon/components/bs-modal.js [hunk 8]
```diff
@@ -193,6 +193,20 @@ export default class Modal extends Component {
...
+   *  `modal-fullscreen-[x]-down` class when using BS breakpoints
+   *   ([x] = 'sm', 'md', 'lg', 'xl', 'xxl').
+   *  
+   * @property fullscreen
+   * @type {string}
+   * @default 'true'
+   * @public
+   */
+  @arg
+  fullscreen = 'true';
```
Comment (jelhan) [inline addon/components/bs-modal.js:208]: This doesn't seem to be used at all as you are not accessing `{{this.fullscreen}}` in the template but only `{{@fullscreen}}`. I think it could be removed safely.

```suggestion
```

File: addon/components/bs-modal.js [hunk 9]
```diff
@@ -193,6 +193,20 @@ export default class Modal extends Component {
   @arg
   scrollable = false;
 
+    /**
+   *  Allows adding fullscreen mode for modals. It will
+   *  apply the `modal-fullscreen' class when using 'true' and
```
Comment (jelhan) [inline addon/components/bs-modal.js:198]: `true` was using quotes instead of backticks and therefore won't be formatted as code.

```suggestion
   *  apply the `modal-fullscreen' class when using `true` and
```

File: addon/components/bs-modal.js [hunk 10]
```diff
@@ -193,6 +193,20 @@ export default class Modal extends Component {
...
+   *  `modal-fullscreen-[x]-down` class when using BS breakpoints
+   *   ([x] = 'sm', 'md', 'lg', 'xl', 'xxl').
+   *  
+   * @property fullscreen
+   * @type {string}
```
Comment (jelhan) [inline addon/components/bs-modal.js:203]: Reading the description I think the type is either `boolean` or `string`.

```suggestion
   * @type {(Boolean|String)}
```

**Review by jelhan — CHANGES_REQUESTED** [review]
Thanks a lot for providing this pull request!

Looks good overall to me. I only have noticed some minor points. Would be great if you could fix them as well.

@simonihmig Would be great if you could have a look at the proposed public API. I haven't found any similar case in our existing source code. Are you fine with having a union type (boolean or string) here?

File: addon/components/bs-modal.js (hunk 2 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:270]: Yes exact!!!

File: addon/components/bs-modal.js (hunk 3 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:281]: Yes exact!!!

File: addon/components/bs-modal/dialog.hbs (hunk 4 above)
Comment (Ayoahha) [reply addon/components/bs-modal/dialog.hbs:23]: OK. Fixed

File: addon/components/bs-modal/dialog.hbs (hunk 5 above)
Comment (Ayoahha) [reply addon/components/bs-modal/dialog.hbs:24]: I think this is my linter or beautifyer addon !

File: tests/integration/components/bs-modal-test.js (hunk 6 above)
Comment (Ayoahha) [reply tests/integration/components/bs-modal-test.js:317]: Fixed

File: addon/components/bs-modal.js (hunk 8 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:208]: Fixed

File: addon/components/bs-modal.js (hunk 9 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:198]: OK, but in this case, some of the other comments have also quotes !
So we need to update all of these ?

File: addon/components/bs-modal.js (hunk 10 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:203]: Fixed

File: tests/integration/components/bs-modal-test.js (hunk 7 above)
Comment (Ayoahha) [reply tests/integration/components/bs-modal-test.js:316]: 1. Fixed
2. I do not understand what you are asking...

File: addon/components/bs-modal.js (hunk 9 above)
Comment (jelhan) [reply addon/components/bs-modal.js:198]: It is okay to only address the new API docs added by this pull request here. I would recommend to fix other issues in separate pull requests to keep focus.

I just noticed that at the end of `modal-fullscreen` it is also a single quote and not a backtick. Here is an updated suggestion.

```suggestion
   *  apply the `modal-fullscreen` class when using `true` and
```

File: tests/integration/components/bs-modal-test.js (hunk 7 above)
Comment (jelhan) [reply tests/integration/components/bs-modal-test.js:316]: The test suite uses `modules` to group related tests. The main module is declared very at the top of the file: https://github.com/kaliber5/ember-bootstrap/blob/e24988a19d7b21607ee7f386b49ce3f9b0159b0c/tests/integration/components/bs-modal-test.js#L11 But it contains an additional module declared later focused on the animation: https://github.com/kaliber5/ember-bootstrap/blob/e24988a19d7b21607ee7f386b49ce3f9b0159b0c/tests/integration/components/bs-modal-test.js#L251 You have put your new test within that module. It would be better if you could move your test out of that sub-module. So basically moving it above line 251.

Comment (Ayoahha) [reply tests/integration/components/bs-modal-test.js]: Exact. Understood
**Fixed**

File: addon/components/bs-modal.js (hunk 9 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js]: Fixed

File: addon/components/bs-modal/dialog.hbs (hunk 4 above)
Comment (jelhan) [reply addon/components/bs-modal/dialog.hbs]: I think you must use inline `{{if}}` in this case. https://guides.emberjs.com/release/components/conditional-content/#toc_inline-if

Comment (Ayoahha) [reply addon/components/bs-modal/dialog.hbs]: Yes it was my proposal... But it seems tricky here in this case...

File: addon/components/bs-modal.js [hunk 11]
```diff
@@ -193,6 +193,18 @@ export default class Modal extends Component {
...
 
+  /**
+   *  Allows adding fullscreen mode for modals. It will
+   *  apply the `modal-fullscreen' class when using `true` and
+   *  `modal-fullscreen-[x]-down` class when using BS breakpoints
+   *   ([x] = 'sm', 'md', 'lg', 'xl', 'xxl').
+   *
+   * @property fullscreen
+   * @type {(Boolean|String)}
+   * @default true
```
Comment (jelhan) [inline addon/components/bs-modal.js:204]: I think the default value is either `undefined` or `false`. If it would be `true`, the modal would be shown if `<BsModal>` is invoked without setting `@fullscreen` argument.

I guess documenting it as `false` provides most clarity for readers. That the default value actually used is `undefined`, seems to be an implementation detail on our side an user does not need to care about.

```suggestion
   * @default false
```

File: tests/integration/components/bs-modal-test.js [hunk 12]
```diff
@@ -312,5 +312,35 @@ module('Integration | Component | bs-modal', function (hooks) {
       await settled();
       assert.dom('.modal').doesNotExist();
     });
+
+    test('it allows to render modal in fullscreen using @fullscreen argument', async function (assert) {
+      this.set('fullscreen', null);
+      await render(hbs`<BsModal>`);
+      assert.dom('.modal-dialog').doesNotHaveClass(/modal-fullscreen/);
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:319]: I like how you solved that with a RegExp. :clap:

File: addon/components/bs-modal.js [hunk 13]
```diff
@@ -193,6 +193,18 @@ export default class Modal extends Component {
   @arg
   scrollable = false;
 
+  /**
+   *  Allows adding fullscreen mode for modals. It will
+   *  apply the `modal-fullscreen' class when using `true` and
+   *  `modal-fullscreen-[x]-down` class when using BS breakpoints
+   *   ([x] = 'sm', 'md', 'lg', 'xl', 'xxl').
```
Comment (jelhan) [inline addon/components/bs-modal.js:200]: Would be great if you could add the link to the Bootstrap documentation for clarity: https://getbootstrap.com/docs/5.1/components/modal/#fullscreen-modal

File: addon/components/bs-modal/dialog.hbs [hunk 14]
```diff
@@ -13,9 +13,8 @@
   {{did-insert this.getOrSetTitleId}}
   {{did-insert this.setInitialFocus}}
 >
-  <div class="modal-dialog {{this.sizeClass}} {{if @centered "modal-dialog-centered"}} {{if @scrollable "modal-dialog-scrollable"}}" role="document">
-    <div class="modal-content"
-     tabindex="-1"
+<div class="modal-dialog {{this.sizeClass}} {{if @centered "modal-dialog-centered"}} {{if @scrollable "modal-dialog-scrollable"}} {{if @fullscreen (if (bs-eq @fullscreen true) "modal-fullscreen" "modal-fullscreen-{{@fullscreen}}-down")}}" role="document">
```
Comment (jelhan) [inline addon/components/bs-modal/dialog.hbs:16]: I just noticed that fullscreen modals are only supported by Bootstrap 5. Ember Bootstrap supports both Bootstrap 4 and 5. We need to exclude this code if Ember Bootstrap is used with Bootstrap 4.

We are using build-time macros provided by Embroider to do so: `{{macroCondition}}` to have dead-code elimination at build-time. `{{macroGetOwnConfig}}` to get the value to decide, which code path should be eliminated. Together it looks like this:

```hbs
{{macroCondition (macroGetOwnConfig "isBS5") "only-included-in-the-build-if-app-is-using-bootstrap-5"}}
```

You can see an example here: https://github.com/kaliber5/ember-bootstrap/blob/4b261eba2d6352eafdade589e4ba2a5a9f3e750a/addon/components/bs-button.hbs#L4

Please also update the API docs to mention that it is only supported for Bootstrap 5.

File: tests/integration/components/bs-modal-test.js [hunk 15]
```diff
@@ -312,5 +312,35 @@ module('Integration | Component | bs-modal', function (hooks) {
       await settled();
       assert.dom('.modal').doesNotExist();
     });
+
+    test('it allows to render modal in fullscreen using @fullscreen argument', async function (assert) {
+      this.set('fullscreen', null);
+      await render(hbs`<BsModal>`);
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:318]: You need to close the `<BsModal>` component. Otherwise it is a syntax error, which breaks the build.

```suggestion
      await render(hbs`<BsModal>Hello World!</BsModal>`);
```

File: tests/integration/components/bs-modal-test.js [hunk 16]
```diff
@@ -312,5 +312,35 @@ module('Integration | Component | bs-modal', function (hooks) {
...
+      assert.dom('.modal-dialog').doesNotHaveClass(/modal-fullscreen/);
+
+      this.set('fullscreen', true);
+      await render(hbs`<BsModal @fullscreen={{this.fullscreen}}>`);
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:322]: No need to re-render again. The updated `{{this.fullscreen}}` value should be picked up automatically.

```suggestion
```

If it is not working, you may need to `await settled()` before to ensure Ember had time to pick up the state change and update the DOM. To be honest I'm still confused when `this.set()` in tests is triggering a sync re-render and when not.

It's the same for the other cases of course.

**Review by jelhan — CHANGES_REQUESTED** [review]
I have done another review. Noticed that I missed some things in the first one. I'm sorry for raising up that many topics again.

File: addon/components/bs-modal.js (hunk 11 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:204]: Fixed

File: addon/components/bs-modal.js (hunk 13 above)
Comment (Ayoahha) [reply addon/components/bs-modal.js:200]: Fixed

File: addon/components/bs-modal/dialog.hbs (hunk 14 above)
Comment (Ayoahha) [reply addon/components/bs-modal/dialog.hbs:16]: Well done !
Fixed in doc and in hbs

File: tests/integration/components/bs-modal-test.js (hunk 15 above)
Comment (Ayoahha) [reply tests/integration/components/bs-modal-test.js:318]: Of course !! My mistake

File: addon/components/bs-modal/dialog.hbs (hunk 5 above)
Comment (jelhan) [reply addon/components/bs-modal/dialog.hbs:17]:
```suggestion
    <div class="modal-content"
      tabindex="-1"
```

File: tests/integration/components/bs-modal-test.js [hunk 17]
```diff
@@ -127,6 +127,31 @@ module('Integration | Component | bs-modal', function (hooks) {
     assert.dom('.modal').hasAttribute('aria-labelledby', modalTitleId);
   });
 
+  test('it allows to render modal in fullscreen using @fullscreen argument', async function (assert) {
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:130]: As that feature is supported in Bootstrap 5 only, we should also only run the test for BS5. We have an utility function called `testBS5` to do so: https://github.com/kaliber5/ember-bootstrap/blob/4b261eba2d6352eafdade589e4ba2a5a9f3e750a/tests/helpers/bootstrap.js#L26-L28 You could find an usage example here: https://github.com/kaliber5/ember-bootstrap/blob/a254a1ff8de05f0cc6fa5d51bbb92d3ad827b5ce/tests/integration/components/bs-button-group-test.js#L54

File: tests/integration/components/bs-modal-test.js [hunk 18]
```diff
@@ -127,6 +127,31 @@ module('Integration | Component | bs-modal', function (hooks) {
...
+    this.set('fullscreen', null);
+    await render(hbs`<BsModal>Hello World!</BsModal>`);
+    assert.dom('.modal-dialog').doesNotHaveClass(/modal-fullscreen/);
+
+    this.set('fullscreen', true);
+    await render(hbs`<BsModal @fullscreen={{this.fullscreen}}>Hello World!</BsModal>`);
```
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:136]: This `await render()` should not be needed.

```suggestion
```

**Review by jelhan — CHANGES_REQUESTED** [review]
I'm sorry for coming back with another set of changes, which I missed in my previous review.

File: tests/integration/components/bs-modal-test.js (hunk 17 above)
Comment (Ayoahha) [reply tests/integration/components/bs-modal-test.js:130]: Fixed

File: addon/components/bs-modal/dialog.hbs [hunk 19]
```diff
@@ -1,28 +1,40 @@
 <div
-  role="dialog"
-  tabindex="-1"
+  role='dialog'
```
Comment (jelhan) [inline addon/components/bs-modal/dialog.hbs:2]: Need to revert the quote style changes in this file. I guess this happened by running Prettier on the file but not having Prettier configured for Handlebars in the repository.

**Comment by jelhan** [comment]
@simonihmig Thanks a lot for merging. I assumed that I had approved long ago. Seems I missed that. Currently in vacation and therefore haven't had the chance to follow up.
